File: eel-core/src/main/kotlin/io/eels/schema/Field.kt
package io.eels.schema

data class Field(val name: String,
                 val type: FieldType = FieldType.String,
                 val nullable: Boolean = true,
                 val precision: Precision = Precision(0),
                 val scale: Scale = Scale(0),
                 val signed: Boolean = false,
                 val arrayType: FieldType? = null, // if an array then the type of the array elements
                 val fields: List<Field> = emptyList(), // if a struct, then the fields of that struct
                 val comment: String? = null) {
  // Creates a lowercase version of this column
  fun toLowerCase(): Field = copy(name = name.toLowerCase())

  companion object {
    fun createStruct(name: String, vararg fields: Field): Field = Field(name, type = FieldType.Struct, fields = fields.asList())
  }
}

enum class FieldType {
  BigInt,
  Binary,
  Boolean,
  Date,
  Decimal,
  Double,
  Float,
  Int,
  Long,
  Short,
  String,
  Struct,
  Timestamp
}

data class Precision(val value: Int)
data class Scale(val value: Int)

File: eel-core/src/main/scala/io/eels/schema/field.scala
package io.eels.schema

case class Field(name: String,
                 `type`: FieldType = FieldType.String,
                 nullable: Boolean = true,
                 precision: Precision = Precision(0),
                 scale: Scale = Scale(0),
                 signed: Boolean = false,
                 arrayType: Option[FieldType] = None, // if an array then the type of the array elements
                 fields: Seq[Field] = Nil, // if a struct, then the fields of that struct
                 partition: Boolean = false,
                 comment: Option[String] = None) {

  // Creates a lowercase version of this column
  def toLowerCase(): Field = copy(name = name.toLowerCase())

  def withComment(comment: String): Field = copy(comment = Some(comment))
  def withNullable(nullable: Boolean): Field = copy(nullable = nullable)
  def withPartition(partition: Boolean) = copy(partition = partition)
}


object Field {
  def createStruct(name: String, first: Field, rest: Field*): Field = createStruct(name, first +: rest)
  def createStruct(name: String, fields: Seq[Field]): Field = Field(name, `type` = FieldType.Struct, fields = fields)
}

case class Precision(value: Int) extends AnyVal
case class Scale(value: Int) extends AnyVal